vote/vote_broadcaster.go

```diff
@@ -0,0 +1,133 @@
+
+import (
+	"fmt"
+	"github.com/bnb-chain/greenfield-challenger/common"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/bnb-chain/greenfield-challenger/config"
+	"github.com/bnb-chain/greenfield-challenger/db/dao"
+	"github.com/bnb-chain/greenfield-challenger/db/model"
+	"github.com/bnb-chain/greenfield-challenger/executor"
+	"github.com/bnb-chain/greenfield-challenger/logging"
+	"github.com/tendermint/tendermint/votepool"
+)
+
+type VoteBroadcaster struct {
+	daoManager      *dao.DaoManager
+	config          *config.Config
+	signer          *VoteSigner
+	executor        *executor.Executor
+	blsPublicKey    []byte
+	cachedLocalVote map[uint64]*votepool.Vote
+	DataProvider
+}
+
+func NewVoteBroadcaster(cfg *config.Config, dao *dao.DaoManager, signer *VoteSigner,
+	executor *executor.Executor, kind DataProvider,
+) *VoteBroadcaster {
+	return &VoteBroadcaster{
+		config:          cfg,
+		daoManager:      dao,
+		signer:          signer,
+		executor:        executor,
+		DataProvider:    kind,
+		cachedLocalVote: nil,
+		blsPublicKey:    getBlsPubKeyFromPrivKeyStr(cfg.VotePoolConfig.BlsPrivateKey),
+	}
+}
+
+func (p *VoteBroadcaster) BroadcastVotesLoop() {
+	p.cachedLocalVote = make(map[uint64]*votepool.Vote, math.MaxInt)
+	broadcastAttempts := 0
+	for {
+		currentHeight := p.executor.GetCachedBlockHeight()
+		// Ask about this function
+		events, err := p.DataProvider.FetchEventsForSelfVote(currentHeight)
+		if err != nil {
+			logging.Logger.Errorf("vote processor failed to fetch unexpired events to collate votes, err=%+v", err.Error())
+			continue
+		}
+		if len(events) == 0 {
+			time.Sleep(RetryInterval)
+			continue
+		}
+
+		for _, event := range events {
+			localVote := p.cachedLocalVote[event.ChallengeId]
+
+			if localVote == nil {
+				localVote, err = p.constructVoteAndSign(event)
+				if err != nil {
+					if strings.Contains(err.Error(), "Duplicate") {
+						logging.Logger.Errorf("[non-blocking error] broadcaster was trying to save a duplicated vote after clearing cache for challengeId: %d, err=%+v", event.ChallengeId, err.Error())
+					} else {
+						logging.Logger.Errorf("broadcaster ran into error trying to construct vote for challengeId: %d, err=%+v", event.ChallengeId, err.Error())
+						continue
+					}
+				}
+				p.cachedLocalVote[event.ChallengeId] = localVote
+			}
+
+			err = p.broadcastForSingleEvent(localVote, event)
+			if err != nil {
+				continue
+			}
+			time.Sleep(50 * time.Millisecond)
+		}
+
+		broadcastAttempts++
+		if broadcastAttempts == common.CacheClearInterval {
+			// Clear cachedLocalVote every N loops, preCheck cannot catch events expired in between iterations
+			p.cachedLocalVote = make(map[uint64]*votepool.Vote, math.MaxInt)
+			broadcastAttempts = 0
+		}
+
+		time.Sleep(RetryInterval)
+	}
+}
+
+func (p *VoteBroadcaster) broadcastForSingleEvent(localVote *votepool.Vote, event *model.Event) error {
+	err := p.preCheck(event)
+	if err != nil {
+		if strings.Contains(err.Error(), "event expired") {
```

Comment:
Can you define an error to replace this string, to make it clearer? `"event expired"`